### PR TITLE
Initial RMA work

### PR DESCRIPTION
**Pull request by ztiffany**
```
Implement internal RMA interface.

1. Implement NIC TX list alloc/free.  Make TX list use dlist_entry.
2. Implement CQ NIC poll list.
3. Progress NICs used by a CQ during fi_cq_read()
4. Add internal RMA interface.
  -allocate fab_req, txd
  -issue GNI RDMA op
  -write libfabric CQ events after GNI TX completion
```

@jswaro @sungeunchoi @hppritcha

### CODE REVIEW

**Comment by lanl-ompi** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.open-mpi.org/job/edison_libfabric_pr/80/
Test PASSed.

File: prov/gni/include/gnix_util.h [hunk 1]
```diff
@@ -71,7 +71,7 @@ static inline void dlist_remove_init(struct dlist_entry *e)
 #define dlist_entry(e, type, member) container_of(e, type, member)
 
 #define dlist_first_entry(h, type, member)				\
-	(dlist_empty(h) ? NULL : dlist_entry((h)->next, type, member))
+	dlist_entry((h)->next, type, member)
```
Comment (sungeunchoi) [inline prov/gni/include/gnix_util.h]: FYI, this is conflict with PR #197 I just committed.  You'll probably need to rebase.

File: prov/gni/src/gnix_cq.c [hunk 2]
```diff
@@ -327,6 +345,50 @@ ssize_t _gnix_cq_add_error(struct gnix_fid_cq *cq, void *op_context,
 	return ret;
 }
 
+int _gnix_cq_poll_nic_add(struct gnix_fid_cq *cq, struct gnix_nic *nic)
+{
+	struct gnix_cq_poll_nic *pnic, *tmp;
+
+	dlist_for_each_safe(&cq->poll_nics, pnic, tmp, list) {
+		if (pnic->nic == nic) {
+			pnic->ref_cnt++;
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_cq.c:360]: I think you need to use `atomic_inc` rather than just ++.

File: prov/gni/src/gnix_cq.c [hunk 3]
```diff
@@ -327,6 +345,50 @@ ssize_t _gnix_cq_add_error(struct gnix_fid_cq *cq, void *op_context,
...
+
+	pnic = malloc(sizeof(struct gnix_cq_poll_nic));
+	if (!pnic) {
+		GNIX_WARN(FI_LOG_CQ, "Failed to add NIC to CQ poll list.\n");
+		return -FI_ENOMEM;
+	}
+
+	/* EP holds a ref count on the NIC */
+	pnic->nic = nic;
+	pnic->ref_cnt = 1;
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_cq.c:375]: And this should use `atomic_initialize`.

File: prov/gni/src/gnix_cq.c [hunk 4]
```diff
@@ -327,6 +345,50 @@ ssize_t _gnix_cq_add_error(struct gnix_fid_cq *cq, void *op_context,
...
+	return FI_SUCCESS;
+}
+
+int _gnix_cq_poll_nic_rem(struct gnix_fid_cq *cq, struct gnix_nic *nic)
+{
+	struct gnix_cq_poll_nic *pnic, *tmp;
+
+	dlist_for_each_safe(&cq->poll_nics, pnic, tmp, list) {
+		if (pnic->nic == nic) {
+			if (!--pnic->ref_cnt) {
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_cq.c:392]: `atomic_dec` here.

File: prov/gni/src/gnix_mr.c [hunk 5]
```diff
@@ -65,10 +65,11 @@ void _gnix_convert_key_to_mhdl(
...
 		INOUT gni_mem_handle_t *mhdl)
 {
-	uint64_t va = (uint64_t) __sign_extend(key->pfn << PAGE_SHIFT,
-			GNIX_MR_VA_BITS);
+	uint64_t va = key->pfn;
 	uint8_t flags = 0;
 
+	va = (uint64_t) __sign_extend(va << PAGE_SHIFT, GNIX_MR_VA_BITS);
+
 	if (key->flags & GNIX_MR_FLAG_READONLY)
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_mr.c:295]: Am I missing something?  What's the reason for this change?

File: prov/gni/src/gnix_nic.c [hunk 6]
```diff
@@ -52,12 +52,231 @@ static pthread_mutex_t gnix_nic_list_lock = PTHREAD_MUTEX_INITIALIZER;
...
+			}
+			gnix_tdesc = container_of(gni_desc,
+						struct gnix_tx_descriptor,
+						desc.gni_desc);
+		}  else if (GNI_CQ_GET_TYPE(cqe)
+			    == GNI_CQ_EVENT_TYPE_SMSG) {
+			msg_id = GNI_CQ_GET_MSG_ID(cqe);
+			gnix_tdesc = gnix_desc_lkup_by_id(nic,msg_id);
+			if (gnix_tdesc == NULL)
+				ret = -FI_ENOENT;
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_nic.c:108]: This return path leaves the function while still holding the lock.  I don't think it's intentional, but if it is, I would strongly suggest a redesign.

File: prov/gni/src/gnix_nic.c [hunk 7]
```diff
@@ -52,12 +52,231 @@ static pthread_mutex_t gnix_nic_list_lock = PTHREAD_MUTEX_INITIALIZER;
...
+		/*
+		 * check whether CQE from SMSG or a Post
+		 * transaction
+		 */
+		if (GNI_CQ_GET_TYPE(cqe)
+				== GNI_CQ_EVENT_TYPE_POST) {
+			status2 = GNI_GetCompleted(nic->tx_cq, cqe, &gni_desc);
+			if ((status2 != GNI_RC_SUCCESS) &&
+				(status2 != GNI_RC_TRANSACTION_ERROR)) {
+				ret = gnixu_to_fi_errno(status2);
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_nic.c:98]: This return value can get lost later in the case.  Is this intentional?

File: prov/gni/src/gnix_nic.c [hunk 8]
```diff
@@ -52,12 +52,231 @@ static pthread_mutex_t gnix_nic_list_lock = PTHREAD_MUTEX_INITIALIZER;
...
+
+	default:
+		assert(0);  /* TODO: better error later */
+	}
+
+	/*
+	 * keep on dequeuing until we get GNI_RC_NOT_DONE
+	 */
+
+	goto try_again;
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_nic.c:171]: Is there any reason to _not_ use a `while(true)` loop here instead of the goto?

File: prov/gni/src/gnix_nic.c [hunk 9]
```diff
@@ -52,12 +52,231 @@ static pthread_mutex_t gnix_nic_list_lock = PTHREAD_MUTEX_INITIALIZER;
...
+	return ret;
+
+}
+
+/*
+ * process an RX CQEs for this nic
+ */
+static int __nic_rx_progress(struct gnix_nic *nic)
+{
+	return -FI_ENOSYS;
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_nic.c:184]: Maybe this should just return FI_SUCCESS, until it's implemented so that `_gnix_nic_progress` succeeds.

File: prov/gni/src/gnix_nic.c [hunk 10]
```diff
@@ -81,14 +300,14 @@ static int __gnix_nic_tx_freelist_init(struct gnix_nic *nic, int n_descs)
...
+	dlist_init(&nic->tx_desc_free_list);
+	dlist_init(&nic->tx_desc_active_list);
 
-	for (i = 0, desc_ptr = desc_base; i < n_descs; i++) {
+	for (i = 0, desc_ptr = desc_base; i < n_descs; i++, desc_ptr++) {
 		desc_ptr->desc.id = i;
-		gnix_list_node_init(&desc_ptr->desc.list);
-		list_add_tail(&nic->tx_desc_free_list,
-				&desc_ptr->desc.list);
+		dlist_init(&desc_ptr->desc.list);
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_nic.c]: I actually don't think this matters, but the corresponding dlist call is `dlist_node_init`.  @hppritcha Is there a reason for making this call?  Seems like ithe pointers are just going to be reset by the call to `dlist_insert_tail` on the next line.  I also left it as is in `_gnix_dgram_hndl_alloc`, but maybe that call should go away also.

File: prov/gni/src/gnix_rma.c [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+		GNIX_INFO(FI_LOG_EP_DATA, "lmdh: %llx:%llx rmdh: %llx:%llx key: %llx\n",
+			  *(uint64_t *)tl_mdh, *(((uint64_t *)tl_mdh) + 1),
+			  *(uint64_t *)tr_mdh, *(((uint64_t *)tr_mdh) + 1),
+			  fab_req->rem_mr_key);
+	}
+
+	status = GNI_PostRdma(fab_req->vc->gni_ep, &txd->desc.gni_desc);
+	if (status != GNI_RC_SUCCESS) {
+		GNIX_INFO(FI_LOG_EP_DATA, "GNI_PostRdma() failed: %d\n", status);
+		return -FI_ENOSPC;
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_rma.c]: Is there a reason to not call `gnixu_to_fi_errno` here?

File: prov/gni/src/gnix_rma.c [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+		return GNI_POST_RDMA_GET;
+	default:
+		break;
+	}
+
+	GNIX_WARN(FI_LOG_EP_DATA, "Unsupported post type: %d", fr_type);
+	return -FI_ENOSYS;
+}
+
+int _gnix_post_req(struct gnix_fab_req *fab_req)
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_rma.c]: Should this be static (and double underscore prefix)?

File: prov/gni/test/rdm_rw.c [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+/*******************************************************************************
+ * Test vc functions.
+ ******************************************************************************/
+
+TestSuite(rdm_rma, .init = rdm_rma_setup, .fini = rdm_rma_teardown,
+	  .disabled = false);
+
+#define BUF_SZ 4096
+Test(rdm_rma, rw)
+{
```
Comment (sungeunchoi) [inline prov/gni/test/rdm_rw.c:185]: NICE!

@ztiffany @hppritcha @jswaro @jshimek 
On a related note, should we have the unit tests silent (no output) by default?  Maybe we could use the criterion verbose flag to get more verbose output?  FWIW, in the automated builds, I use `--verbose`, so we would see the output if there is an error.

File: prov/gni/include/gnix_util.h (hunk 1 above)
Comment (ztiffany) [inline prov/gni/include/gnix_util.h]: Will do.

File: prov/gni/src/gnix_cq.c (hunk 2 above)
Comment (ztiffany) [inline prov/gni/src/gnix_cq.c:360]: I need to go back and consider locking everywhere.  I may wrap uses of cq->poll_nics with a CQ lock.

File: prov/gni/src/gnix_mr.c (hunk 5 above)
Comment (ztiffany) [inline prov/gni/src/gnix_mr.c:295]: The pfn bitfield isn't wide enough to hold the shifted va value.  The result va was being truncated.

Comment (sungeunchoi) [inline prov/gni/src/gnix_mr.c:295]: Ah thanks for the clarification.  Can you add a comment so that no one changes it back?

File: prov/gni/src/gnix_nic.c (hunk 9 above)
Comment (ztiffany) [inline prov/gni/src/gnix_nic.c:184]: The bottom half of _gnix_nic_progress isn't debugged yet.  Making this change causes crashes.  More work is needed on the progress stuff.  I need to sync up with Howard to see if he's developed any more.

File: prov/gni/src/gnix_rma.c (hunk 12 above)
Comment (ztiffany) [inline prov/gni/src/gnix_rma.c]: We may need to expose it later when sending queued requests.  I'll hide it for now

File: prov/gni/src/gnix_nic.c (hunk 10 above)
Comment (ztiffany) [inline prov/gni/src/gnix_nic.c]: I'll remove that.

File: prov/gni/src/gnix_nic.c [hunk 14]
```diff
@@ -52,12 +52,231 @@ static pthread_mutex_t gnix_nic_list_lock = PTHREAD_MUTEX_INITIALIZER;
...
+		/* this shouldn't happen SMSG */
+		if (GNI_CQ_GET_TYPE(cqe) == GNI_CQ_EVENT_TYPE_SMSG) {
+			char ebuf[512];
+			GNI_CqErrorStr(cqe, ebuf, sizeof(ebuf));
+			GNIX_WARN(FI_LOG_EP_DATA,
+				  "CQ error statusfor GNI_SmsgSend - %s\n",
+				  ebuf);
+			goto err1;
+		}
+		status2 = GNI_GetCompleted(nic->tx_cq, cqe, &gni_desc);
```
Comment (jswaro) [inline prov/gni/src/gnix_nic.c:138]: There should be a more meaningful name than status2

File: prov/gni/src/gnix_rma.c [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+	txd->desc.gni_desc.dlvr_mode = GNI_DLVMODE_PERFORMANCE; /* check flags */
+	txd->desc.gni_desc.local_addr = (uint64_t)fab_req->loc_addr;
+	txd->desc.gni_desc.local_mem_hndl = md->mem_hndl;
+	txd->desc.gni_desc.remote_addr = (uint64_t)fab_req->rem_addr;
+	txd->desc.gni_desc.remote_mem_hndl = mdh;
+	txd->desc.gni_desc.length = fab_req->len;
+	txd->desc.gni_desc.rdma_mode = 0; /* check flags */
+	txd->desc.gni_desc.src_cq_hndl = fab_req->gnix_ep->nic->tx_cq; /* check flags */
+
+	{
```
Comment (jswaro) [inline prov/gni/src/gnix_rma.c:135]: This comes down to style, but I don't like using internal scoping to achieve things that should belong in an static/inline function. Could you push this out to another function?

File: prov/gni/src/gnix_nic.c (hunk 6 above)
Comment (hppritcha) [inline prov/gni/src/gnix_nic.c:108]: which case would return from `___nic_tx_progress` without releasing the lock?

File: prov/gni/src/gnix_nic.c (hunk 7 above)
Comment (hppritcha) [inline prov/gni/src/gnix_nic.c:98]: no i'll fix this one.

File: prov/gni/src/gnix_nic.c (hunk 8 above)
Comment (hppritcha) [inline prov/gni/src/gnix_nic.c:171]: I did this to avoid another tab indention.

File: prov/gni/src/gnix_nic.c (hunk 9 above)
Comment (hppritcha) [inline prov/gni/src/gnix_nic.c:184]: @ztiffany lets do that today (6/2) looks like from MPI forum agenda that afternoon would work well - 1 - 3:30 CDT.

Comment (hppritcha) [inline prov/gni/src/gnix_nic.c:184]: I've coded up the rx side and will merge that work in to the #185.

File: prov/gni/src/gnix_nic.c (hunk 14 above)
Comment (hppritcha) [inline prov/gni/src/gnix_nic.c:138]: changed to status_gc

File: prov/gni/src/gnix_nic.c (hunk 6 above)
Comment (sungeunchoi) [inline prov/gni/src/gnix_nic.c:108]: Sorry, I think I was wrong.  I'll blame it on Friday.

**Comment by lanl-ompi** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.open-mpi.org/job/edison_libfabric_pr/100/
Test PASSed.

File: prov/gni/src/gnix_rma.c (hunk 15 above)
Comment (ztiffany) [inline prov/gni/src/gnix_rma.c:135]: It's just 2 debug print statements that I'm not quite ready to erase.  I'm just using the block to keep the debug specific variables next to the print statements.

**Comment by ztiffany** [comment]
I updated the code.  The rdm_rw criterion test now exits successfully.  That's due to closing MRs in the test and also updates to gnix_cq.c.

gnix_cq.c now explicitly prevents rebinding an EP to a new CQ.  Also, fi_close(ep) removes ref counts from it's CQs,

**Comment by ztiffany** [comment]
I probably need to add some locking still.  I'm guessing:
1. lock the NIC during GNI_RdmaPost()
2. lock the NIC during progress?
3. lock CQ poll list accesses

**Comment by hppritcha** [comment]
yes these need locks

2015-06-02 12:36 GMT-06:00 ztiffany notifications@github.com:

**Comment by lanl-ompi** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.open-mpi.org/job/edison_libfabric_pr/101/
Test PASSed.

**Comment by ztiffany** [comment]
I added locks:
1. nic->lock protects the TX freelist and calls to GNI_PostRdma
2. cq->nic_lock rwlock protects cq->poll_nics list
3. ep->lock protects the fabric request freelist

File: prov/gni/include/gnix.h [hunk 16]
```diff
@@ -339,19 +340,26 @@ struct gnix_fab_req {
...
+
+	/* RMA stuff */
+	uint64_t loc_addr;
+	void *loc_md;
+	uint64_t rem_addr;
+	uint64_t rem_mr_key;
+	size_t len;
+	uint64_t flags;
+	uint64_t data;
+	uint64_t tag;
```
Comment (hppritcha) [inline prov/gni/include/gnix.h:362]: Not for this commit, but eventually we'll probably need to figure out how to compact this structure.  Its getting pretty big.

File: prov/gni/src/gnix_cq.c [hunk 17]
```diff
@@ -247,6 +248,27 @@ static struct slist_entry *alloc_cq_entry(size_t size)
...
+		if (rc) {
+			GNIX_WARN(FI_LOG_CQ,
+				  "_gnix_nic_progress failed: %d\n", rc);
+		}
+	}
+
+	rwlock_unlock(&cq->nic_lock);
+
+	return FI_SUCCESS;
+}
```
Comment (hppritcha) [inline prov/gni/src/gnix_cq.c:270]: I'd rather this break out of the loop over polled nics and return rc rather than FI_SUCCESS in the case that `_gnix_nic_progress` returns an error.

Comment (ztiffany) [inline prov/gni/src/gnix_cq.c:270]: Do you want that change in this initial version?  Is that the last of the comments?

Comment (hppritcha) [inline prov/gni/src/gnix_cq.c:270]: answers:
no
yes

2015-06-02 16:01 GMT-06:00 ztiffany notifications@github.com:

**Comment by lanl-ompi** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.open-mpi.org/job/edison_libfabric_pr/104/
Test PASSed.
